src/main/scala/chisel3/interface/Interface.scala

```diff
@@ -0,0 +1,121 @@
+// SPDX-License-Identifier: Apache-2.0
+package chisel3.interface
+
+import chisel3.{BlackBox => _, Module => _, _}
+import chisel3.experimental.{BaseModule, FlatIO}
+import chisel3.experimental.dataview._
+import scala.annotation.implicitNotFound
+
+@implicitNotFound(
+  "this method requires information from the separable compilation implementation, please bring one into scope as an `implicit val`. You can also consult the team that owns the implementation to refer to which one you should use!"
+)
+trait ConformsTo[Intf <: Interface, Mod <: BaseModule] {
+
+  /** Return the module that conforms to a port-level interface. */
+  private[interface] def genModule(): Mod
+
+  /** Define how this module hooks up to the port-level interface. */
+  private[interface] def portMap: Seq[(Mod, Intf#Ports) => (Data, Data)]
+
+}
+
+/** Functionality which is common */
+sealed trait InterfaceCommon {
+
+  private[interface] type Ports <: Record
+
+  /** Returns the Record that is the port-level interface. */
+  private[interface] def ports(): Ports
+
+}
+
+/** A generator of different Interfaces. Currently, this is just an Interface
+  * that is not a Singleton.
+  */
+trait InterfaceGenerator extends InterfaceCommon
+
+/** An interface between hardware units. Any module that implements this
+  * interface may be separately compiled from any module that instantiates this
+  * interface.
+  */
+trait Interface extends InterfaceCommon { self: Singleton =>
+
+  /** This types represents the type of a valid conformance to this Interface.
+    */
+  private type Conformance[Mod <: RawModule] = ConformsTo[this.type, Mod]
```

Comment:
For working with `ExtModule`, I needed to change `RawModule` to `BaseModule` here and below

Reply:
Good catch! Fixed in https://github.com/chipsalliance/chisel/pull/3199/commits/b8a5fc0a2a8c0e455147143388fdb96d33593d0e.